src/harness/unittests/tsserverProjectSystem.ts

```diff
@@ -254,33 +254,34 @@ namespace ts.projectSystem {
     }
 
     export function isFolder(s: FSEntry): s is Folder {
-        return isArray((<Folder>s).entries);
+        return s && isArray((<Folder>s).entries);
     }
 
     export function isFile(s: FSEntry): s is File {
-        return typeof (<File>s).content === "string";
+        return s && typeof (<File>s).content === "string";
     }
 
-    export function addFolder(fullPath: string, toPath: (s: string) => Path, fs: Map<FSEntry>): Folder {
-        const path = toPath(fullPath);
-        if (fs.has(path)) {
-            Debug.assert(isFolder(fs.get(path)));
-            return (<Folder>fs.get(path));
+    function invokeDirectoryWatcher(callbacks: DirectoryWatcherCallback[], getRelativeFilePath: () => string) {
+        if (callbacks) {
+            const cbs = callbacks.slice();
+            for (const cb of cbs) {
+                const fileName = getRelativeFilePath();
+                cb(fileName);
+            }
         }
+    }
 
-        const entry: Folder = { path, entries: [], fullPath };
-        fs.set(path, entry);
-
-        const baseFullPath = getDirectoryPath(fullPath);
-        if (fullPath !== baseFullPath) {
-            addFolder(baseFullPath, toPath, fs).entries.push(entry);
+    function invokeFileWatcher(callbacks: FileWatcherCallback[], fileName: string, eventId: FileWatcherEventKind) {
+        if (callbacks) {
+            const cbs = callbacks.slice();
+            for (const cb of cbs) {
+                cb(fileName, eventId);
+            }
         }
-
-        return entry;
     }
 
     export function checkMapKeys(caption: string, map: Map<any>, expectedKeys: string[]) {
-        assert.equal(map.size, expectedKeys.length, `${caption}: incorrect size of map`);

```

Comment:
this allows the caller to have expectedKeys in any order and thats great help in test since one doesnt have to rely on the order when writing test